Add scope option to the route subcommand

robgc opened the pull request:
This PR solves #80.

### Changes

- Added `-s` | `--scope <SCOPE>` option to `npc`. Allowing users to filter routes according the scope provided.

### Notes

- In order to use the `possible_values` function when defining a `clap::Arg` the `routes` subcommand had to be defined outside of the main definition of the CLI because the `clap_app!` macro syntax does not have support for that function. This function is used to perform validation of the values passed to the argument and to expand the information provided in the `--help`.

cathay4t reviewed (changes requested):
* Please use `possible_values` of clap for this. https://docs.rs/clap/2.33.3/clap/struct.Arg.html#method.possible_values
 * Please also support full name of scope like `global` or `link`. Try to use the same name `ip -d route` is using.

robgc commented:
- In 19caa63 I use `possible_values` to limit the values that can be passed to the `--scope` option. I do not know if you wish me to use it in another flag or subcommand.
- In 20f9a81 I added full names as supported values, including the ones from `ip route` that we were missing.

ffmancera commented:
Can you update the existing PR? The CI has been updated. Thanks!

robgc commented:
@ffmancera I have updated the PR. Now CI tasks should be performed.

cathay4t commented on src/cli/npc.rs:
```diff
@@ -99,6 +106,29 @@ fn _is_route_to_specified_dev(route: &Route, iface_name: &str) -> bool {
+        || route_scope_str == ROUTE_SCOPE_STRINGS[3]
+    {
+        RouteScope::Universe
+    } else if route_scope_str == ROUTE_SCOPE_STRINGS[4]
+        || route_scope_str == ROUTE_SCOPE_STRINGS[5]
+    {
+        RouteScope::Site
+    } else if route_scope_str == ROUTE_SCOPE_STRINGS[6]
+        || route_scope_str == ROUTE_SCOPE_STRINGS[7]
+    {
+        RouteScope::Link
+    } else if route_scope_str == ROUTE_SCOPE_STRINGS[8]
+        || route_scope_str == ROUTE_SCOPE_STRINGS[9]
```
Hard coded magic number is vulnerable to human errors.

How about:

```rust
match route_scope_str {
    "universe" || "u" => RouteScope::Universe,
   _ => Err("Invalid route scope"),
}
```

And the filter line could be:

```rust
.filter(|route| scope == "all" || route.scope == get_route_scope_from_str(scope))
```

robgc replied:
Thanks for the suggestions! I followed them and I have added some code to handle the error case when the route scope is invalid, it returns a `CliError` if necessary.

I kept the `ROUTE_SCOPE_STRINGS` array so `clap` can validate the users' input beforehand.

cathay4t commented:
@robgc Sorry for longggggg wait, I have amended your patch with better approach.

cathay4t committed:
npc: Add scope option to the route subcommand

This patch is based on the work of Roberto Garcia Calero.

Add command line argument: `npc route --scope`:
 * a|all: no filter
 * g|global|u|universe: `RouteScope::Universe`
 * s|site: `RouteScope::Site`
 * l|link: `RouteScope::Link`
 * h|host: `RouteScope::Host`
 * n|nowhere|no_where: `RouteScope::NoWhere`

By default, `npc route` show routes from all scope.

Implemented `From<&str> ` and `Display` trait for RouteScope.

Co-authored-by: Roberto Garcia Calero <9429166+robgc@users.noreply.github.com>

Signed-off-by: Gris Ge <fge@redhat.com>
